// SPDX-FileCopyrightText: 2025 Mathis Logemann <mathis@quite.rocks>
//
// SPDX-License-Identifier: MIT

syntax = "proto3";
package quite.proto;

// If the value is a non queryable object, use the metatypes
message ClassValue {
  string type_name = 1;

  message ClassMember {
    string name = 1;
    Value value = 2;
  };
  repeated ClassMember value = 2;
}

// If the value is a queryable object (RemoteObject) just return the object_id.
message ObjectValue {
  uint64 object_id = 1;
}

message ArrayValue {
  repeated Value value = 1;
}

message Value {
  uint64 type_id = 1;
  oneof value_oneof {
    string string_val = 2;
    int64 int_val = 3;
    uint64 uint_val = 4;
    double double_val = 5;
    bool bool_val = 6;
    ClassValue class_val = 7;
    ObjectValue object_val = 8;
    ArrayValue array_val = 9;
  }
}
